// vidstack/dist/prod/chunks/chunk-N224OYCW.js
import { preconnect } from './chunk-IVW6GMZV.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_create_walker, $$_effect, $$_attr, $$_listen, $$_create_template } from 'maverick.js/dom';
import { signal, computed, effect } from 'maverick.js';
import { defineElement, Component } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><img part="img" />`);
var Poster = class extends Component {
  constructor() {
    super(...arguments);
    this.Gb = signal(true);
    this.Yc = signal(false);
  }
  onAttach(el) {
    this.a = useMedia();
    this.Zc = computed(this.Wh.bind(this));
    this.ue = this.Xh.bind(this);
    this.setAttributes({
      "data-loading": this.Gb,
      "aria-hidden": $ariaBool(this.z.bind(this))
    });
  }
  onConnect(el) {
    const { canLoad, poster } = this.a.$store;
    window.requestAnimationFrame(() => {
      if (!canLoad())
        preconnect(poster());
    });
    effect(this.ia.bind(this));
  }
  z() {
    const { poster } = this.a.$store;
    return this.Yc() || !poster();
  }
  Wh() {
    const { canLoad, poster } = this.a.$store;
    return canLoad() && poster().length ? poster() : null;
  }
  Xh() {
    return this.Zc() ? this.$props.alt() : null;
  }
  ia() {
    const { canLoad, poster } = this.a.$store;
    const isLoading = canLoad() && !!poster();
    this.Gb.set(isLoading);
    this.Yc.set(false);
  }
  Yh() {
    this.Gb.set(false);
  }
  ja() {
    this.Gb.set(false);
    this.Yc.set(true);
  }
  render() {
    const { crossorigin } = this.a.$store;
    return (() => {
      const [$$_root, $$_walker] = $$_create_walker($$_templ);
      $$_effect(() => $$_attr($$_root, "src", this.Zc()));
      $$_effect(() => $$_attr($$_root, "alt", this.ue()));
      $$_effect(() => $$_attr($$_root, "crossorigin", crossorigin()));
      $$_listen($$_root, "load", this.Yh.bind(this));
      $$_listen($$_root, "error", this.ja.bind(this));
      return $$_root;
    })();
  }
};
Poster.el = defineElement({
  tagName: "media-poster",
  props: { alt: void 0 }
});

export { Poster };
