// vidstack/dist/prod/chunks/chunk-4VMKL2VS.js
import { toggleButtonProps, ToggleButton } from './chunk-S6JBW74X.js';
import { setARIALabel } from './chunk-HQQTSFXD.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { Icon } from './chunk-OSS74BPE.js';
import { volume_high_default } from './chunk-KQUCP2EG.js';
import { volume_low_default } from './chunk-DZHDFPQR.js';
import { mute_default } from './chunk-IVWVADOA.js';
import { $$_create_component } from 'maverick.js/dom';
import { defineElement } from 'maverick.js/element';

var MuteButton = class extends ToggleButton {
  constructor() {
    super(...arguments);
    this.na = "toggleMuted";
  }
  onAttach(el) {
    this.a = useMedia();
    this.v = this.V.bind(this);
    setARIALabel(el, this.G.bind(this));
    this.setAttributes({
      "data-muted": this.v,
      "data-volume": this.ug.bind(this)
    });
    super.onAttach(el);
  }
  H(event) {
    const remote = this.a.remote;
    this.v() ? remote.unmute(event) : remote.mute(event);
  }
  V() {
    const { muted, volume } = this.a.$store;
    return muted() || volume() === 0;
  }
  G() {
    return this.v() ? "Unmute" : "Mute";
  }
  ug() {
    const { muted, volume } = this.a.$store, $volume = volume();
    if (muted() || $volume === 0)
      return "muted";
    else if ($volume >= 0.5)
      return "high";
    else if ($volume < 0.5)
      return "low";
  }
  render() {
    return [
      $$_create_component(Icon, { paths: volume_high_default, slot: "volume-high" }),
      $$_create_component(Icon, { paths: volume_low_default, slot: "volume-low" }),
      $$_create_component(Icon, { paths: mute_default, slot: "volume-muted" })
    ];
  }
};
MuteButton.el = defineElement({
  tagName: "media-mute-button",
  props: toggleButtonProps
});

export { MuteButton };
