// vidstack/dist/prod/chunks/chunk-EOOVGDNO.js
import { renderRadioGroup } from './chunk-KRJ7Q224.js';
import { ClassManager } from './chunk-H7N2HN5V.js';
import { RadioGroup } from './chunk-7QRO24S6.js';
import { Radio } from './chunk-HGWWXG2M.js';
import { MenuItems } from './chunk-N4YPMUQV.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_create_walker, $$_insert_at_marker_lite, $$_create_template } from 'maverick.js/dom';
import { effect } from 'maverick.js';
import { defineElement } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><span part="label"><!$></span>`);
var AudioMenuItems = class extends MenuItems {
  constructor(instance) {
    super(instance);
    this.a = useMedia();
  }
  onConnect(el) {
    effect(this.ua.bind(this));
    effect(this.va.bind(this));
    const { radioClass, radioCheckClass } = this.$props;
    new ClassManager(el).u("media-radio", radioClass).u('[part="check"]', radioCheckClass);
  }
  va() {
    const { emptyLabel } = this.$props, { audioTrack } = this.a.$store, track = audioTrack();
    this.A.hb.set(track?.label ?? emptyLabel());
  }
  ua() {
    this.A.sa(this.p());
  }
  p() {
    const { audioTracks } = this.a.$store;
    return audioTracks().length === 0;
  }
  o(event) {
    if (this.p())
      return;
    const radioGroup = event.target;
    const value = radioGroup.value, index = this.a.audioTracks.toArray().findIndex((track) => track.label.toLowerCase() === value);
    if (index >= 0)
      this.a.remote.changeAudioTrack(index, event);
  }
  Z() {
    const { audioTrack } = this.a.$store;
    const track = audioTrack();
    return track ? track.label.toLowerCase() : "";
  }
  _() {
    const { audioTracks } = this.a.$store;
    return audioTracks().map((track) => ({
      value: track.label.toLowerCase(),
      content: () => (() => {
        const [$$_root, $$_walker] = $$_create_walker($$_templ), $$_expr = $$_walker.nextNode();
        $$_insert_at_marker_lite($$_expr, track.label);
        return $$_root;
      })()
    }));
  }
  render() {
    const { radioGroupClass } = this.$props;
    return renderRadioGroup({
      value: this.Z.bind(this),
      options: this._.bind(this),
      radioGroupClass,
      onChange: this.o.bind(this)
    });
  }
};
AudioMenuItems.el = defineElement({
  tagName: "media-audio-menu-items",
  props: {
    emptyLabel: "Default",
    radioClass: null,
    radioGroupClass: null,
    radioCheckClass: null
  }
});
AudioMenuItems.register = [RadioGroup, Radio];

export { AudioMenuItems };
