// vidstack/dist/prod/chunks/chunk-DTE723BW.js
import { isTrackCaptionKind } from './chunk-IIULWXWS.js';
import { toggleButtonProps, ToggleButton } from './chunk-S6JBW74X.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { setARIALabel } from './chunk-HQQTSFXD.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { Icon } from './chunk-OSS74BPE.js';
import { closed_captions_on_default } from './chunk-FJ2OEY6P.js';
import { closed_captions_default } from './chunk-3XSNY5HY.js';
import { $$_create_component } from 'maverick.js/dom';
import { defineElement } from 'maverick.js/element';

var CaptionButton = class extends ToggleButton {
  constructor() {
    super(...arguments);
    this.na = "toggleCaptions";
  }
  onAttach(el) {
    this.a = useMedia();
    this.v = this.V.bind(this);
    super.onAttach(el);
    setARIALabel(el, this.G.bind(this));
    this.setAttributes({
      "aria-hidden": $ariaBool(this.z.bind(this))
    });
  }
  H(event) {
    this.a.remote.toggleCaptions(event);
  }
  V() {
    const { textTrack } = this.a.$store, track = textTrack();
    return !!track && isTrackCaptionKind(track);
  }
  z() {
    const { textTracks } = this.a.$store;
    return textTracks().filter(isTrackCaptionKind).length == 0;
  }
  G() {
    const { textTrack } = this.a.$store;
    return textTrack() ? "Closed-Captions Off" : "Closed-Captions On";
  }
  render() {
    return [
      $$_create_component(Icon, { paths: closed_captions_on_default, slot: "on" }),
      $$_create_component(Icon, { paths: closed_captions_default, slot: "off" })
    ];
  }
};
CaptionButton.el = defineElement({
  tagName: "media-caption-button",
  props: toggleButtonProps
});

export { CaptionButton };
