// vidstack/dist/prod/chunks/chunk-7QRO24S6.js
import { radioGroupContext } from './chunk-4L7R5OAB.js';
import { setAttributeIfEmpty } from './chunk-HQQTSFXD.js';
import { menuContext } from './chunk-HSL7QSWM.js';
import { __decorateClass } from './chunk-VPSTUDLO.js';
import { signal, provideContext, hasProvidedContext, effect, peek } from 'maverick.js';
import { defineElement, prop, Component } from 'maverick.js/element';

var RadioGroup = class extends Component {
  constructor(instance) {
    super(instance);
    this.Oa = /* @__PURE__ */ new Set();
    this.Y = signal("");
    this.Fh = this.o.bind(this);
    provideContext(radioGroupContext, {
      add: this.Dh.bind(this),
      remove: this.Eh.bind(this)
    });
  }
  get values() {
    return Array.from(this.Oa).map((radio) => radio.Y());
  }
  get value() {
    return this.Y();
  }
  set value(value) {
    this.o(value);
  }
  onAttach(el) {
    const isMenuItem = hasProvidedContext(menuContext);
    if (!isMenuItem)
      setAttributeIfEmpty(el, "role", "radiogroup");
    this.ne();
    this.setAttributes({ value: this.Y });
  }
  onConnect() {
    effect(this.ne.bind(this));
  }
  onDestroy() {
    this.Oa.clear();
  }
  Dh(radio) {
    if (this.Oa.has(radio))
      return;
    this.Oa.add(radio);
    radio.Fb = this.Fh;
    radio.ib(radio.Y() === this.Y());
  }
  Eh(radio) {
    radio.Fb = null;
    this.Oa.delete(radio);
  }
  ne() {
    this.o(this.$props.value());
  }
  o(newValue, trigger) {
    const currentValue = peek(this.Y);
    if (!newValue || newValue === currentValue)
      return;
    const currentRadio = this.oe(currentValue), newRadio = this.oe(newValue);
    currentRadio?.ib(false, trigger);
    newRadio?.ib(true, trigger);
    this.Y.set(newValue);
    this.dispatch("change", { trigger });
  }
  oe(newValue) {
    for (const radio of this.Oa) {
      if (newValue === peek(radio.Y))
        return radio;
    }
    return null;
  }
};
RadioGroup.el = defineElement({
  tagName: "media-radio-group",
  props: { value: "" }
});
__decorateClass([
  prop
], RadioGroup.prototype, "values", 1);
__decorateClass([
  prop
], RadioGroup.prototype, "value", 1);

export { RadioGroup };
