// vidstack/dist/prod/chunks/chunk-J6NAGVT3.js
import { toggleButtonProps, ToggleButton } from './chunk-S6JBW74X.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { setARIALabel } from './chunk-HQQTSFXD.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { Icon } from './chunk-OSS74BPE.js';
import { picture_in_picture_exit_default } from './chunk-U6WW63HQ.js';
import { picture_in_picture_default } from './chunk-KMLSRSZ2.js';
import { $$_create_component } from 'maverick.js/dom';
import { defineElement } from 'maverick.js/element';

var PIPButton = class extends ToggleButton {
  constructor() {
    super(...arguments);
    this.na = "togglePictureInPicture";
  }
  onAttach(el) {
    this.a = useMedia();
    this.v = this.V.bind(this);
    super.onAttach(el);
    setARIALabel(el, this.G.bind(this));
    const { pictureInPicture } = this.a.$store;
    this.setAttributes({
      "aria-hidden": $ariaBool(this.z.bind(this)),
      "data-pip": pictureInPicture
    });
  }
  H(event) {
    const remote = this.a.remote;
    this.v() ? remote.exitPictureInPicture(event) : remote.enterPictureInPicture(event);
  }
  V() {
    const { pictureInPicture } = this.a.$store;
    return pictureInPicture();
  }
  z() {
    const { canPictureInPicture } = this.a.$store;
    return !canPictureInPicture();
  }
  G() {
    const { pictureInPicture } = this.a.$store;
    return pictureInPicture() ? "Exit Picture In Picture" : "Enter Picture In Picture";
  }
  render() {
    return [
      $$_create_component(Icon, { paths: picture_in_picture_default, slot: "enter" }),
      $$_create_component(Icon, { paths: picture_in_picture_exit_default, slot: "exit" })
    ];
  }
};
PIPButton.el = defineElement({
  tagName: "media-pip-button",
  props: toggleButtonProps
});

export { PIPButton };
