// vidstack/dist/prod/chunks/chunk-PIHXPUOQ.js
import { SliderStoreFactory } from './chunk-53B74NMI.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_create_walker, $$_effect, $$_attr, $$_listen, $$_ref, $$_create_template } from 'maverick.js/dom';
import { signal, useStore, computed, effect } from 'maverick.js';
import { defineElement, Component } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><video muted="" playsinline="" preload="auto" part="video" style="max-width: unset"></video>`);
var SliderVideo = class extends Component {
  constructor() {
    super(...arguments);
    this.k = null;
    this.db = signal(false);
    this.Dc = signal(false);
  }
  onAttach() {
    this.a = useMedia();
    this.I = useStore(SliderStoreFactory);
    this.da = computed(this.Tg.bind(this));
    this.Ec = computed(this.z.bind(this));
    this.setAttributes({
      "data-loading": this.Bc.bind(this),
      "aria-hidden": $ariaBool(this.Ec)
    });
    effect(this.Ug.bind(this));
    effect(this.Rb.bind(this));
  }
  onConnect() {
    if (this.k.readyState >= 2)
      this.ob();
  }
  render() {
    const { crossorigin } = this.a.$store;
    return (() => {
      const [$$_root, $$_walker] = $$_create_walker($$_templ);
      $$_effect(() => $$_attr($$_root, "src", this.da()));
      $$_effect(() => $$_attr($$_root, "crossorigin", crossorigin()));
      $$_listen($$_root, "canplay", this.ob.bind(this));
      $$_listen($$_root, "error", this.ja.bind(this));
      $$_ref($$_root, this.Vg.bind(this));
      return $$_root;
    })();
  }
  Tg() {
    const { canLoad } = this.a.$store;
    return canLoad() ? this.$props.src() : null;
  }
  Bc() {
    return !this.db() && !this.Ec();
  }
  z() {
    const { duration } = this.a.$store;
    return !!this.Dc() || !this.db() || !Number.isFinite(duration());
  }
  Ug() {
    this.da();
    this.db.set(false);
    this.Dc.set(false);
  }
  ob(event) {
    this.db.set(true);
    this.dispatch("can-play", { trigger: event });
  }
  ja(event) {
    this.Dc.set(true);
    this.dispatch("error", { trigger: event });
  }
  Rb() {
    const { duration } = this.a.$store;
    const { pointerRate } = this.I;
    if (this.db() && this.k && Number.isFinite(duration()) && Number.isFinite(pointerRate())) {
      this.k.currentTime = pointerRate() * duration();
    }
  }
  Vg(el) {
    this.k = el;
  }
};
SliderVideo.el = defineElement({
  tagName: "media-slider-video",
  props: { src: void 0 }
});

export { SliderVideo };
