// vidstack/dist/prod/chunks/chunk-VHYXNS3B.js
import { isTrackCaptionKind } from './chunk-IIULWXWS.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { effect, peek } from 'maverick.js';
import { defineElement, Component } from 'maverick.js/element';
import { listenEvent, createDisposalBin } from 'maverick.js/std';
import { CaptionsRenderer, renderVTTCueString, updateTimedVTTCueNodes } from 'media-captions';

var CaptionsTextRenderer = class {
  constructor(_renderer) {
    this.M = _renderer;
    this.priority = 10;
    this.m = null;
    this.$ = createDisposalBin();
  }
  attach() {
  }
  canRender() {
    return true;
  }
  detach() {
    this.$.empty();
    this.M.reset();
    this.m = null;
  }
  changeTrack(track) {
    if (!track || this.m === track)
      return;
    this.$.empty();
    if (track.readyState < 2) {
      this.M.reset();
      this.$.add(
        listenEvent(track, "load", () => this.te(track), { once: true })
      );
    } else {
      this.te(track);
    }
    this.$.add(
      listenEvent(track, "add-cue", (event) => {
        this.M.addCue(event.detail);
      }),
      listenEvent(track, "remove-cue", (event) => {
        this.M.removeCue(event.detail);
      })
    );
    this.m = track;
  }
  te(track) {
    this.M.changeTrack({
      cues: [...track.cues],
      regions: [...track.regions]
    });
  }
};

// src/player/ui/captions/captions.tsx
var Captions = class extends Component {
  onAttach() {
    this.a = useMedia();
    this.setAttributes({
      "aria-hidden": $ariaBool(this.z.bind(this))
    });
  }
  onConnect(el) {
    this.M = new CaptionsRenderer(el);
    this.xa = new CaptionsTextRenderer(this.M);
    effect(this.Nh.bind(this));
  }
  onDisconnect() {
    if (this.xa) {
      this.xa.detach();
      this.a.textRenderers.remove(this.xa);
    }
    this.M?.destroy();
  }
  z() {
    const { textTrack } = this.a.$store, track = textTrack();
    return !track || !isTrackCaptionKind(track);
  }
  Nh() {
    const { viewType } = this.a.$store;
    if (viewType() === "audio") {
      return this.Oh();
    } else {
      return this.Ph();
    }
  }
  Oh() {
    effect(this.Lc.bind(this));
    return () => {
      this.el.textContent = "";
    };
  }
  Lc() {
    if (this.z())
      return;
    const { textTrack } = this.a.$store;
    listenEvent(textTrack(), "cue-change", this.Qh.bind(this));
    effect(this.Rh.bind(this));
  }
  Qh() {
    this.el.textContent = "";
    const { currentTime, textTrack } = this.a.$store, time = peek(currentTime), activeCues = peek(textTrack).activeCues;
    for (const cue of activeCues) {
      const cueEl = document.createElement("div");
      cueEl.setAttribute("part", "cue");
      cueEl.innerHTML = renderVTTCueString(cue, time);
      this.el.append(cueEl);
    }
  }
  Rh() {
    const { currentTime } = this.a.$store;
    updateTimedVTTCueNodes(this.el, currentTime());
  }
  Ph() {
    effect(this.Sh.bind(this));
    effect(this.Th.bind(this));
    this.a.textRenderers.add(this.xa);
    return () => {
      this.el.textContent = "";
      this.xa.detach();
      this.a.textRenderers.remove(this.xa);
    };
  }
  Sh() {
    this.M.dir = this.$props.textDir();
  }
  Th() {
    if (this.z())
      return;
    const { currentTime } = this.a.$store;
    this.M.currentTime = currentTime();
  }
};
Captions.el = defineElement({
  tagName: "media-captions",
  props: { textDir: "ltr" }
});

export { Captions };
