// vidstack/dist/prod/chunks/chunk-CMYZQZ7P.js
import { getNumberOfDecimalPlaces } from './chunk-CVLY5S52.js';
import { isMediaStream, isHLSSrc } from './chunk-FRNAFKFA.js';
import { IS_SAFARI } from './chunk-WSYKLWQE.js';
import { isUndefined, isNumber, setAttribute, isString, useDisposalBin, listenEvent, DOMEvent, isNil } from 'maverick.js/std';
import { effect, onDispose } from 'maverick.js';

var RAFLoop = class {
  constructor(_callback) {
    this.ha = _callback;
  }
  mb() {
    if (!isUndefined(this.za))
      return;
    this.dd();
  }
  ga() {
    if (isNumber(this.za))
      window.cancelAnimationFrame(this.za);
    this.za = void 0;
  }
  dd() {
    this.za = window.requestAnimationFrame(() => {
      if (isUndefined(this.za))
        return;
      this.ha();
      this.dd();
    });
  }
};

// src/player/core/providers/html/html–media-events.ts
var HTMLMediaEvents = class {
  constructor(_provider, _context) {
    this.j = _provider;
    this.c = _context;
    this.$ = useDisposalBin();
    this.Ua = false;
    this.ed = false;
    this.fd = false;
    this.nb = new RAFLoop(this.Ce.bind(this));
    this.Xe = void 0;
    this.fi = void 0;
    this.De();
    effect(this.Ee.bind(this));
    onDispose(this.Ob.bind(this));
  }
  get a() {
    return this.j.media;
  }
  get g() {
    return this.c.delegate;
  }
  Ob() {
    this.nb.ga();
    this.$.empty();
  }
  /**
   * The `timeupdate` event fires surprisingly infrequently during playback, meaning your progress
   * bar (or whatever else is synced to the currentTime) moves in a choppy fashion. This helps
   * resolve that by retrieving time updates in a request animation frame loop.
   */
  Ce() {
    const newTime = this.j.currentTime;
    if (this.c.$store.currentTime() !== newTime)
      this.Aa(newTime);
  }
  De() {
    this.r("loadstart", this.ia);
    this.r("abort", this.gd);
    this.r("emptied", this.Fe);
    this.r("error", this.ja);
  }
  Ge() {
    if (this.ed)
      return;
    this.$.add(
      this.r("loadeddata", this.He),
      this.r("loadedmetadata", this.Ie),
      this.r("canplay", this.ob),
      this.r("canplaythrough", this.Je),
      this.r("durationchange", this.Ke),
      this.r("play", this.Le),
      this.r("progress", this.Me),
      this.r("stalled", this.Ne),
      this.r("suspend", this.Oe)
    );
    this.ed = true;
  }
  Pe() {
    if (this.fd)
      return;
    this.$.add(
      this.r("pause", this.Qe),
      this.r("playing", this.Re),
      this.r("ratechange", this.Se),
      this.r("seeked", this.Te),
      this.r("seeking", this.Ue),
      this.r("ended", this.Ve),
      this.r("volumechange", this.Pb),
      this.r("waiting", this.We)
    );
    this.fd = true;
  }
  r(eventType, handler) {
    return listenEvent(
      this.a,
      eventType,
      handler.bind(this)
    );
  }
  gi(event2) {
    return;
  }
  Aa(time, trigger) {
    this.g.d("time-update", {
      // Avoid errors where `currentTime` can have higher precision.
      detail: {
        currentTime: Math.min(time, this.c.$store.seekableEnd()),
        played: this.a.played
      },
      trigger
    });
  }
  ia(event2) {
    if (this.a.networkState === 3) {
      this.gd(event2);
      return;
    }
    this.Ge();
    this.g.d("load-start", { trigger: event2 });
  }
  gd(event2) {
    this.g.d("abort", { trigger: event2 });
  }
  Fe() {
    this.g.d("emptied", { trigger: event });
  }
  He(event2) {
    this.g.d("loaded-data", { trigger: event2 });
  }
  Ie(event2) {
    this.hd();
    this.Pe();
    this.g.d("volume-change", {
      detail: {
        volume: this.a.volume,
        muted: this.a.muted
      }
    });
    this.g.d("loaded-metadata", { trigger: event2 });
    if (IS_SAFARI && isHLSSrc(this.c.$store.source())) {
      this.g.jd(this.Qb(), event2);
    }
  }
  Qb() {
    return {
      duration: this.a.duration,
      buffered: this.a.buffered,
      seekable: this.a.seekable
    };
  }
  hd() {
    const isLive = !Number.isFinite(this.a.duration);
    this.g.d("stream-type-change", {
      detail: isLive ? "live" : "on-demand"
    });
  }
  Le(event2) {
    if (!this.c.$store.canPlay)
      return;
    this.g.d("play", { trigger: event2 });
  }
  Qe(event2) {
    if (this.a.readyState === 1 && !this.Ua)
      return;
    this.Ua = false;
    this.nb.ga();
    this.g.d("pause", { trigger: event2 });
  }
  ob(event2) {
    this.g.jd(this.Qb(), event2);
  }
  Je(event2) {
    if (this.c.$store.started())
      return;
    this.g.d("can-play-through", {
      trigger: event2,
      detail: this.Qb()
    });
  }
  Re(event2) {
    this.Ua = false;
    this.g.d("playing", { trigger: event2 });
    this.nb.mb();
  }
  Ne(event2) {
    this.g.d("stalled", { trigger: event2 });
    if (this.a.readyState < 3) {
      this.Ua = true;
      this.g.d("waiting", { trigger: event2 });
    }
  }
  We(event2) {
    if (this.a.readyState < 3) {
      this.Ua = true;
      this.g.d("waiting", { trigger: event2 });
    }
  }
  Ve(event2) {
    this.nb.ga();
    this.Aa(this.a.duration, event2);
    this.g.d("end", { trigger: event2 });
    if (this.c.$store.loop()) {
      this.Ye();
    } else {
      this.g.d("ended", { trigger: event2 });
    }
  }
  Ee() {
    if (this.c.$store.paused()) {
      listenEvent(this.a, "timeupdate", this.Rb.bind(this));
    }
  }
  Rb(event2) {
    this.Aa(this.a.currentTime, event2);
  }
  Ke(event2) {
    this.hd();
    if (this.c.$store.ended()) {
      this.Aa(this.a.duration, event2);
    }
    this.g.d("duration-change", {
      detail: this.a.duration,
      trigger: event2
    });
  }
  Pb(event2) {
    this.g.d("volume-change", {
      detail: {
        volume: this.a.volume,
        muted: this.a.muted
      },
      trigger: event2
    });
  }
  Te(event2) {
    this.Aa(this.a.currentTime, event2);
    this.g.d("seeked", {
      detail: this.a.currentTime,
      trigger: event2
    });
    if (Math.trunc(this.a.currentTime) === Math.trunc(this.a.duration) && getNumberOfDecimalPlaces(this.a.duration) > getNumberOfDecimalPlaces(this.a.currentTime)) {
      this.Aa(this.a.duration, event2);
      if (!this.a.ended) {
        this.c.player.dispatchEvent(
          new DOMEvent("media-play-request", {
            trigger: event2
          })
        );
      }
    }
  }
  Ue(event2) {
    this.g.d("seeking", {
      detail: this.a.currentTime,
      trigger: event2
    });
  }
  Me(event2) {
    this.g.d("progress", {
      detail: {
        buffered: this.a.buffered,
        seekable: this.a.seekable
      },
      trigger: event2
    });
  }
  Ye() {
    const hasCustomControls = isNil(this.a.controls);
    if (hasCustomControls)
      this.a.controls = false;
    this.c.player.dispatchEvent(new DOMEvent("media-loop-request"));
  }
  Oe(event2) {
    this.g.d("suspend", { trigger: event2 });
  }
  Se(event2) {
    this.g.d("rate-change", {
      detail: this.a.playbackRate,
      trigger: event2
    });
  }
  ja(event2) {
    const error = this.a.error;
    if (!error)
      return;
    this.g.d("error", {
      detail: {
        message: error.message,
        code: error.code,
        mediaError: error
      },
      trigger: event2
    });
  }
};

// src/foundation/list/symbols.ts
var LIST_ADD = Symbol(0);
var LIST_REMOVE = Symbol(0);
var LIST_RESET = Symbol(0);
var LIST_SELECT = Symbol(0);
var LIST_READONLY = Symbol(0);
var LIST_SET_READONLY = Symbol(0);
var LIST_ON_RESET = Symbol(0);
var LIST_ON_REMOVE = Symbol(0);
var LIST_ON_USER_SELECT = Symbol(0);

// src/player/core/providers/html/native-audio-tracks.ts
var NativeAudioTracks = class {
  constructor(_provider, _context) {
    this.j = _provider;
    this.c = _context;
    this.Va.onaddtrack = this.Ze.bind(this);
    this.Va.onremovetrack = this._e.bind(this);
    this.Va.onchange = this.$e.bind(this);
    listenEvent(this.c.audioTracks, "change", this.af.bind(this));
  }
  get Va() {
    return this.j.media.audioTracks;
  }
  Ze(event2) {
    const _track = event2.track;
    if (_track.label === "")
      return;
    const audioTrack = {
      id: _track.id + "",
      label: _track.label,
      language: _track.language,
      kind: _track.kind,
      selected: false
    };
    this.c.audioTracks[LIST_ADD](audioTrack, event2);
    if (_track.enabled)
      audioTrack.selected = true;
  }
  _e(event2) {
    const track = this.c.audioTracks.getById(event2.track.id);
    if (track)
      this.c.audioTracks[LIST_REMOVE](track, event2);
  }
  $e(event2) {
    let enabledTrack = this.kd();
    if (!enabledTrack)
      return;
    const track = this.c.audioTracks.getById(enabledTrack.id);
    if (track)
      this.c.audioTracks[LIST_SELECT](track, true, event2);
  }
  kd() {
    return Array.from(this.Va).find((track) => track.enabled);
  }
  af(event2) {
    const { current } = event2.detail;
    if (!current)
      return;
    const track = this.Va.getTrackById(current.id);
    if (track) {
      const prev = this.kd();
      if (prev)
        prev.enabled = false;
      track.enabled = true;
    }
  }
};

// src/player/core/providers/html/provider.ts
var HTMLMediaProvider = class {
  constructor(_media) {
    this.a = _media;
  }
  setup(context) {
    new HTMLMediaEvents(this, context);
    if ("audioTracks" in this.media)
      new NativeAudioTracks(this, context);
  }
  get type() {
    return "";
  }
  get media() {
    return this.a;
  }
  get paused() {
    return this.a.paused;
  }
  get muted() {
    return this.a.muted;
  }
  set muted(muted) {
    this.a.muted = muted;
  }
  get volume() {
    return this.a.volume;
  }
  set volume(volume) {
    this.a.volume = volume;
  }
  get currentTime() {
    return this.a.currentTime;
  }
  set currentTime(time) {
    this.a.currentTime = time;
  }
  get playsinline() {
    return this.a.hasAttribute("playsinline");
  }
  set playsinline(playsinline) {
    setAttribute(this.a, "playsinline", playsinline);
  }
  get playbackRate() {
    return this.a.playbackRate;
  }
  set playbackRate(rate) {
    this.a.playbackRate = rate;
  }
  async play() {
    return this.a.play();
  }
  async pause() {
    return this.a.pause();
  }
  async loadSource({ src }, preload) {
    this.a.preload = preload;
    if (isMediaStream(src)) {
      this.a.srcObject = src;
    } else {
      this.a.srcObject = null;
      this.a.src = isString(src) ? src : window.URL.createObjectURL(src);
    }
    this.a.load();
  }
};

export { HTMLMediaProvider, LIST_ADD, LIST_ON_REMOVE, LIST_ON_RESET, LIST_ON_USER_SELECT, LIST_READONLY, LIST_REMOVE, LIST_RESET, LIST_SELECT, LIST_SET_READONLY, RAFLoop };
