// vidstack/dist/prod/chunks/chunk-HGWWXG2M.js
import { radioGroupContext } from './chunk-4L7R5OAB.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { onPress } from './chunk-HQQTSFXD.js';
import { menuContext } from './chunk-HSL7QSWM.js';
import { FocusVisibleController } from './chunk-VGM6UASR.js';
import { __decorateClass } from './chunk-VPSTUDLO.js';
import { $$_next_template, $$_create_template } from 'maverick.js/dom';
import { signal, effect, hasProvidedContext, useContext, peek } from 'maverick.js';
import { defineElement, prop, Component } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><div part="check"></div>`);
var Radio = class extends Component {
  constructor(instance) {
    super(instance);
    this.ta = signal(false);
    this.D = {
      Y: this.$props.value,
      ib: this.ib.bind(this),
      Fb: null
    };
    new FocusVisibleController(instance);
  }
  get checked() {
    return this.ta();
  }
  onAttach() {
    effect(this.vc.bind(this));
    this.pe();
    const isMenuItem = hasProvidedContext(menuContext);
    this.setAttributes({
      value: this.$props.value,
      role: isMenuItem ? "menuitemradio" : "radio",
      tabindex: isMenuItem ? -1 : 0,
      "aria-checked": $ariaBool(this.ta)
    });
  }
  onConnect(el) {
    this.pe();
    onPress(el, this.H.bind(this));
  }
  onDisconnect() {
    const group = useContext(radioGroupContext);
    group.remove(this.D);
  }
  pe() {
    const group = useContext(radioGroupContext);
    group.add(this.D);
  }
  vc() {
    const { value } = this.$props, newValue = value();
    if (peek(this.ta)) {
      this.D.Fb?.(newValue);
    }
  }
  H(event) {
    if (peek(this.ta))
      return;
    this.ta.set(true);
    this.dispatch("change", { trigger: event });
    this.D.Fb?.(peek(this.$props.value), event);
  }
  ib(value, trigger) {
    if (peek(this.ta) === value)
      return;
    this.ta.set(value);
    this.dispatch("change", { trigger });
  }
  render() {
    return $$_next_template($$_templ);
  }
};
Radio.el = defineElement({
  tagName: "media-radio",
  props: { value: "" }
});
__decorateClass([
  prop
], Radio.prototype, "checked", 1);

export { Radio };
